src/strat_engine/blockdev.rs

```diff
@@ -0,0 +1,373 @@
+            dbgp!("{} header CRC failed", path.display());
+            return Err(EngineError::Io(io::Error::new(ErrorKind::InvalidInput,
+                                                      format!("{} header CRC failed",
+                                                              path.display()))));
+            // TODO: Try to read end-of-disk copy
+        }
+
+        let stratis_id = from_utf8(&buf[32..64]).unwrap();
+        let id = from_utf8(&buf[64..96]).unwrap();
+
+        let mda_size = Sectors(LittleEndian::read_u32(&buf[160..164]) as u64);
+
+        if *mda_size % 2 != 0 {
+            return Err(EngineError::Io(io::Error::new(ErrorKind::InvalidInput,
+                                                      format!("{} mda size is not an even \
+                                                               number",
+                                                              path.display()))));
+        }
+
+        if mda_size < MIN_MDA_SIZE {
+            return Err(EngineError::Io(io::Error::new(ErrorKind::InvalidInput,
+                                                      format!("{} mda size is less than \
+                                                               minimum ({})",
+                                                              path.display(),
+                                                              *MIN_MDA_SIZE))));
+        }
+
+        Ok(BlockDev {
+            stratis_id: stratis_id.to_owned(),
+            id: id.to_owned(),
+            dev: dev,
+            path: path.to_owned(),
+            sectors: Sectors(try!(blkdev_size(&f)) / SECTOR_SIZE),
+            mdaa: MDA {
+                last_updated: Timespec::new(LittleEndian::read_u64(&buf[64..72]) as i64,
+                                            LittleEndian::read_u32(&buf[72..76]) as i32),
+                used: LittleEndian::read_u32(&buf[76..80]),
+                length: (*mda_size / 2 * SECTOR_SIZE) as u32,
+                crc: LittleEndian::read_u32(&buf[80..84]),
+                offset: SectorOffset(0),
+            },
+            mdab: MDA {
+                last_updated: Timespec::new(LittleEndian::read_u64(&buf[96..104]) as i64,
+                                            LittleEndian::read_u32(&buf[104..108]) as i32),
+                used: LittleEndian::read_u32(&buf[108..112]),
+                length: (*mda_size / 2 * SECTOR_SIZE) as u32,
+                crc: LittleEndian::read_u32(&buf[112..116]),
+                offset: SectorOffset(*mda_size / 2),
+            },
+            mda_sectors: Sectors(LittleEndian::read_u32(&buf[160..164]) as u64),
+            reserved_sectors: Sectors(LittleEndian::read_u32(&buf[164..168]) as u64),
+        })
+    }
+
+    pub fn to_save(&self) -> BlockDevSave {
+        BlockDevSave {
+            path: self.path.clone(),
+            sectors: self.sectors,
+        }
+    }
+
+    pub fn find_all() -> EngineResult<Vec<BlockDev>> {
```

Comment:
Any function that opens and reads every device node is bound to make me jumpy. But maybe that's what everybody does, sanctioned by long practice?